### PR TITLE
Update index.md  indexEnd is not the end since we actually end before it

### PR DESCRIPTION
**Pull request by Stevepurpose**
For long the slice has been a source of confusion among developers. Apart from it's similarity with the splice in terms of name , I feel the naming of the parameters has also contributed. Also I had to specify what happens when we have negative arguments because clearly what happens with single negative arguments is different from what happens generally with the slice. Also naming our second parameter indexEnd is confusing because that particular index is not a part of it , as long as we counting from 0.

### Description

Had to change parameter names and threw more light on negative indexes.

### Motivation

The slice has been a source of confusion for so long and I think the parameter names and negative indexes contributed.

### CODE REVIEW

**Review by Josh-Cena — CHANGES_REQUESTED** [review]
Please revert the parameter name changes. I don't think they make things much clearer, but such long variable names are not conventional in JavaScript. The current names are definitely not wrong: `indexEnd` does not mean "the last index of the slice", but "an index indicating where to end the slice", and whether that particular index is included should be further explained by the prose.

Once you reverted the naming, from there we can discuss the extra content changes.

File: files/en-us/web/javascript/reference/global_objects/string/slice/index.md [hunk 1]
```diff
@@ -63,34 +63,85 @@ console.log(str5); // ""
 
 ### Using slice() with negative indexes
 
-The following example uses `slice()` with negative indexes.
+The slice method with negative indexes counts backwards from left to right.The following example uses `slice()` with negative indexes.
+
+## Syntax
```
Comment (Josh-Cena) [inline files/en-us/web/javascript/reference/global_objects/string/slice/index.md:68]: Why is there another syntax box, and why are you using h2 headers in an h3 section?

We only have one syntax section per page. A parameter can be multi-purpose but that's either documented in the parameter description or the description body.

File: files/en-us/web/javascript/reference/global_objects/string/slice/index.md [hunk 2]
```diff
@@ -63,34 +63,85 @@ console.log(str5); // ""
...
+
+```js-nolint
+slice(showIndexBefore)
+slice(ignoreIndexFrom,ignoreIndexBefore)
+```
+
+## Single negative index
```
Comment (Josh-Cena) [inline files/en-us/web/javascript/reference/global_objects/string/slice/index.md:75]: No need for these extra headings: everything is already under the "Using negative indexes" umbrella heading. This is the structure convention for examples.

**Review by Josh-Cena — COMMENTED** [review]
Structure also needs a lot more work, but we can discuss that later.

**Comment by Stevepurpose** [comment]
I am a bit confused because it seems my explanations are not understood
correctly. As far as negative indexes are concerned ,much explanations are
not given on how the count is done and it is confusing to a beginner.you
might get it because you are an experienced developer. I thought the
purpose of documentation was for explanation.With the previous explanations
particularly on negative index ,it is not clear enough.

**Comment by Stevepurpose** [comment]
Talking about indexStart and indexEnd ,the name is confusing also because
in English ,the end of something is a part of it, you can't refer to
something as an end if it is not a part of it..If our index count was from
1 and not 0 then we could consider it the indexEnd.. I also read somewhere
that JavaScript variable names have no limit. Infact most tutorials
encourage the use of longer variable names to make things better
understood.If structure is the issue I can work on the things you want me
to work on but as far as variable names I have seen longer.

**Comment by Josh-Cena** [comment]
It is okay to use longer variables in a real project, but for syntax blocks we use one or two words. It is impossible to get the complete meaning through one name. For example, it's not strictly speaking an "index" because otherwise negative numbers won't work. We often use even less clear words like "number" and "target", because really the variable names are just there to provide a name we can refer to later. Documentation provides information through prose, not through a variable name.

> in English ,the end of something is a part of it, you can't refer to something as an end if it is not a part of it

`indexEnd` means "where to end extraction", not "where the end of the slice is".

> As far as negative indexes are concerned ,much explanations are not given on how the count is done and it is confusing to a beginner.

I am okay with the extra examples provided; I am asking for the naming change to be reverted and I can take a look at the rest. There are also structural and formatting issues to be worked on, such as the automated reviews you already got above.

**Comment by Stevepurpose** [comment]
Does that mean you want to stick to indexStart and indexEnd?...if length is
the problem,  How do you see ignoreBefore and ignoreFrom?

**Comment by Josh-Cena** [comment]
I still don't like it very much—people tend to think of slice as "where the start is, where the end is", and then remember the boundary conditions; few think of slice as "before are ignored, after are ignored".

**Comment by Stevepurpose** [comment]
The confusion I have seen with people generally is with the splice and
slice and it is because of that start word, but with ignoreBefore and the
prose explanation they know the new string starts from that point since
indexes before that have technically been cut off and indexEnd generally
isn't the end of the old string or new string, the index before the
indexEnd is the end while we cut off from that point hence ignoreFrom since
that's where our cut starts as we know what ignore means and JavaScript as
a higher level programming language being close to the human language .

**Comment by Stevepurpose** [comment]
Kindly take a look at this link to an article I recently published to get
my point of view on the slice.. This article was basically about explaining
concepts and reason for naming my parameters how I did with each explanation

https://dev.to/stevepurpose/the-slice-philosophy-3jgi

**Comment by Josh-Cena** [comment]
I'll ask the rest of the team for opinions. For now I am still opposed although not totally unconvinced. Please understand that this naming change significantly changes how we usually name our variables.

**Comment by Stevepurpose** [comment]
Totally understand!  It is also applicable to array slice of you don't mind
can I submit a PR for that too?

**Comment by Josh-Cena** [comment]
Once we get this to a mergeable state, you should update this PR so that `Array.prototype.slice` and `TypedArray.prototype.slice` are both updated accordingly, yes.

**Comment by Stevepurpose** [comment]
Ok. Thanks

**Comment by hamishwillee** [comment]
I am very strongly opposed to this change. 

The original text is clear and unambiguous English, while the new text is odd - perhaps because referring to a returned range in terms of "what is ignored around it" is unintuitive.
I am convinced these changes actually make the behaviour harder to understand. 

It would be "nice" to come up with a parameter name that somehow captures that the first index is inclusive while the second is non-inclusive of the returned range, but I can't think of one. The text description of the parameter does make this very clear though.

I would not accept the changes before the Example section.  The  changes in the examples need subediting, but this is perhaps the best way to clear up any confusion. 

I would consider using syntax highlighting on include/exclude to make the issue more obvious and perhaps some words after the text on the end option.

```
- `indexStart`
  - : The index of the first character to _include_ in the returned substring.
- `indexEnd` {{optional_inline}}
  - : The index of the first character to _exclude_ from the returned substring (the element at this value is _not_ returned in the range).
```

**Comment by OnkarRuikar** [comment]
Suggested `ignoreIndexBefore-ignoreIndexFrom` are not making it simpler than existing names, because we have to read the description to understand them.

> It would be "nice" to come up with a parameter name that somehow captures that the first index is inclusive while the second is non-inclusive of the returned range, but I can't think of one. The text description of the parameter does make this very clear though.

How about `inclusiveStart-exclusiveEnd`, `inclusiveStart-nonInclusiveEnd`, `indexStart-indexExclusiveEnd`, `start-exclusiveEnd`, or `startIndex-exclusiveEndIndex`?

The current description is also cryptic ('the first character to'). We can have more descriptive text. How about following?
```md
- `indexStart`
  - : Slicing operation starts before this index. Returned slice starts with the character at this index.
- `exclusiveEndIndex` {{optional_inline}}
  - : Slicing operation ends before this index. Returned slice doesn't include the character at this index.
 ```

**Comment by Stevepurpose** [comment]
Good day Hamish. The fact is that the behavior of the slice has a different
behavior  when dealing with a single negative parameter (you must have
observed) ,how do we marry that ? It is not spelt out .And then using
indexStart , same way you used start for splice and Array.prototype.splice
makes the whole thing entirely more confusing. I have been on  tech forums
online  where devs mix up the splice and slice and when you look closely
you would see the start word has a part.. let's look at it this way, the
start in splice is actually the particular the start index where we start
our operations, but is that really the case for the slice?? no.plus every
other thing I pointed out. I think I pointed all these out. I think the
purpose of documentation is to spell things out  in depth. Kindly take out
20 mins of your time to go through this breakdown I did in this article,
probably you'd get why I did these. Link below 👇
https://dev.to/stevepurpose/the-slice-philosophy-3jgi

**Comment by Stevepurpose** [comment]
Hello onkar , I think you have a good point but if we look at it this way ,
ignore is a very basic word that anyone can understand. If length is an
issue like Joshua initially pointed out ,how about ignoreBefore and
ignoreFrom.
kindly check out the meaning of the word ignore in your oxford dictionary
you'd see that it is very straightforward in whichever context they used it
, but exclusive has different meanings and contexts and JavaScript being a
higher level programming language should be easily understood by humans.

**Comment by Josh-Cena** [comment]
> The current description is also cryptic ('the first character to').

The current text is fully accurate. If "the first character to exclude" is confusing for you, we could say "the first character to be excluded". I prefer this wording to describing the boundary as "inclusive" or "exclusive", because the latter is an extra information to memorize while the former bakes it in very naturally.

**Comment by hamishwillee** [comment]
@OnkarRuikar I don't love these either. It's dislike it less than describing things in terms of what is ignored but as @Stevepurpose says, exclusive is an overloaded term. It is more like `indexStartInclusive`-`indexEndNonInclusive`.

@Stevepurpose We're going to have to agree to disagree. This naming convention is not better than what it is attempting to replace.

I'm supportive of spelling things out in the examples, but unless the current text in the syntax/parameters part is wrong, I'd leave it as it is.

Perhaps @wbamberg might hold another opinion when he is back on Friday.

**Comment by Stevepurpose** [comment]
Ok let's look at it from the angle of a  newbie dev who just finished
looking at the docs learning array methods . He finishes studying the slice
and starts studying the splice and sees the  word "start" used again. That
would be confusing. Also don't forget the negative parameter which I
pointed out , particularly the single negative index ,you would see that
the value it returns is inconsistent with the other slice parameters. So
taking all these into consideration ,I think the slice is more ambiguous
than was spelt out in the documentation and I think that has to be spelt
out.

**Comment by Stevepurpose** [comment]
I still have not submitted the PR to the Array.prototype.slice because it's
basically the same principle and we are yet to clarify this string
issue.words like ignoreBefore, cutOffBefore, chopOffBefore, ignoreFrom,
cutOffFrom or chopOffFrom are very explicit and convey the message with
less prose attached. Whichever you settle for. But in whatever ,don't
forget the behavior of the single negative slice argument.

**Comment by wbamberg** [comment]
> @Stevepurpose We're going to have to agree to disagree. This naming convention is not better than what it is attempting to replace. Perhaps @wbamberg might hold another opinion when he is back on Friday.
> 

Since you asked... I think `slice()` is a tricky API to explain in words, and there really isn't any magic set of words that will. make it easy. And I don't think this PR makes the API easier to understand. The bullet list in [Description](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice#description) is hard to parse but that is the cost of being very precise here.

I do think this means that examples are critical to understanding it, and I do agree that adding an example with a single negative parameter is helpful.

Also, though, in [Using slice() with negative indexes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice#using_slice_with_negative_indexes), the source is like this:

````
The following example uses `slice()` with negative indexes.

```js
const str = "The morning is upon us.";
str.slice(-3); // 'us.'
str.slice(-3, -1); // 'us'
str.slice(0, -1); // 'The morning is upon us'
str.slice(4, -1); // 'morning is upon us'
```

This example counts backwards from the end of the string by `11` to find the
start index and forwards from the start of the string by `16` to find the end
index.

```js
console.log(str.slice(-11, 16)); // "is u"
```

Here it counts forwards from the start by `11` to find the start index and
backwards from the end by `7` to find the end index.

```js
console.log(str.slice(11, -7)); // " is u"
```

These arguments count backwards from the end by `5` to find the start index
and backwards from the end by `1` to find the end index.

```js
console.log(str.slice(-5, -1)); // "n us"
```

````

So first we get a generic example that doesn't explain what it is doing, then a series of examples showing specific aspects (positive/negative, negative/positive, negative/negative) each of which get a paragraph of explanation.

One thing I found hard was understanding which example the explanation attaches to. For example here:

````
The following example uses `slice()` with negative indexes.

```js
const str = "The morning is upon us.";
str.slice(-3); // 'us.'
str.slice(-3, -1); // 'us'
str.slice(0, -1); // 'The morning is upon us'
str.slice(4, -1); // 'morning is upon us'
```

This example counts backwards from the end of the string by `11` to find the
start index and forwards from the start of the string by `16` to find the end
index.

```js
console.log(str.slice(-11, 16)); // "is u"
```

Here it counts forwards from the start by `11` to find the start index and
backwards from the end by `7` to find the end index.

```js
console.log(str.slice(11, -7)); // " is u"
```
````

...which code block does "This example" refer to? Which code block does "Here" refer to? 

Also, the source string "The morning is upon us." is only declared in the first code block, so when I get to the end, I lose the connection with it, and the connection with the source string is critical to understanding these examples.

So personally, I would:
- remove the first, general, example
- make all the other code blocks self-contained
- clearly indicate which explanation attaches to which code block
- add an example with a single negative argument (which is in this PR)

For instance:

````

The following example provides a negative value for `indexStart` and a positive value for `indexEnd`. It counts backwards from the end of the string by `11` to find the
start index and forwards from the start of the string by `16` to find the end
index.

```js
const str = "The morning is upon us.";
console.log(str.slice(-11, 16)); // "is u"
```

The following example provides a negative value for `indexStart` and a positive value for `indexEnd`. It counts forwards from the start by `11` to find the start index and
backwards from the end by `7` to find the end index.

```js
const str = "The morning is upon us.";
console.log(str.slice(11, -7)); // " is u"
```

The following example provides negative values for both `indexStart` and `indexEnd`. It counts backwards from the end by `5` to find the start index and backwards from the end by `1` to find the end index.

```js
const str = "The morning is upon us.";
console.log(str.slice(-5, -1)); // "n us"
```

The following example provides a single negative value. This is interpreted as a value for `indexStart`, so it counts backwards from the end by `7` to find the start index, and includes all the source string from the start index to the end.

```js
const str = "The morning is upon us.";
console.log(str.slice(-7, -1)); // "pon us."
```
````

**Comment by wbamberg** [comment]
The other thing that occurred to me is that a diagram showing the index positions referenced by `indexStart` and `indexEnd` might help?

```
              indexStart        indexEnd
                  ↓               ↓
| 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 |
| T | h | e |   | m | o | r | n | i | n |
                 _______________
                      ↑
                    Result
```

**Comment by OnkarRuikar** [comment]
An image of this would be more reliable as spaces may not render the same width in every user agent.

**Comment by OnkarRuikar** [comment]
An image of this would be more reliable as spaces may not render the same width in every user agent.

**Comment by Josh-Cena** [comment]
> as spaces may not render the same width in every user agent.

This would be inside a code block, so it's monospace.

**Comment by OnkarRuikar** [comment]
> This would be inside a code block, so it's monospace.

:sweat_smile: Real image would be more neat than the ASCII art.

**Comment by Stevepurpose** [comment]
@wbamberg whatever you think is best. indexStart is ok but indexEnd is not
what it is I think that has to change .for the single negative index ,what
do you think?

**Comment by Stevepurpose** [comment]
suggested to change indexEnd are indexSnip,
indexchopFrom,indexCutFromindexIgnoreFrom, ignoreFrom

**Comment by hamishwillee** [comment]
FWIW I think 
- an image is a great idea.
- The  reason I like index is because it implies a _position_ not a _value_, so it implies you might not be taking the value at endIndex. It might be even clearer to say `startPoint` and `endPoint`.
- Yes we need to cover the single value case. I think that much of the information currently in return value should either be in the parameters or duplicated. That's the point at which you need to know that the value can be negative, and what happens if the end value is not specified.
- Lots of examples too.

**Comment by Josh-Cena** [comment]
Folks have almost unanimously agreed to not change the name. In fact, exclusive end index is almost a convention in programming: you have the same behavior in Python, Java, R, etc. The idea is that `slice(index, index + sliceLength)` as well as `slice(0, arr.length)` works as expected, and if you read every reference doc you will find this parameter called `end` or something similar. We'd better not fight against what the field has settled on. You are free to hold your own opinions, though. If you want to add a diagram illustrating the exclusiveness of the end index, you may raise another PR.

**Comment by Stevepurpose** [comment]
I've forgotten about this long ago..Thought you closed it.
